src/mlpack/methods/ne/cne.hpp

```diff
@@ -0,0 +1,257 @@
+   */
+  static void CrossoverWeights(Genome& momGenome, 
+                               Genome& dadGenome, 
+                               Genome& child1Genome, 
+                               Genome& child2Genome)
+  {
+    child1Genome = momGenome;
+    child2Genome = dadGenome;
+    for (int i = 0; i < momGenome.linkGenes.size(); ++i)
+    { // assume genome are the same structure.
+      double t = mlpack::math::RandNormal();
+      if (t > 0)
+      {  // prob = 0.5
+        child1Genome.linkGenes[i].Weight(momGenome.linkGenes[i].Weight());
+        child2Genome.linkGenes[i].Weight(dadGenome.linkGenes[i].Weight());
+      } else
+      {
+        child1Genome.linkGenes[i].Weight(dadGenome.linkGenes[i].Weight());
+        child2Genome.linkGenes[i].Weight(momGenome.linkGenes[i].Weight());
+      }
+    }
+  }
+
+  /** 
+   * Create initial species from seed genome and species size.
+   */
+  void InitSpecies()
+  {
+    species = Species(seedGenome, speciesSize);
+  }
+
+  /**
+   * Reproduce the next species. Heart function for CNE.
+   *
+   * Select parents from previous generation based on genomes' fitness.
+   * Apply search operators to parents and produce offspring which
+   * form next generation.
+   */
+  void Reproduce()
+  {
+    // Sort species by fitness
+    species.SortGenomes();
+
+    // Select parents from elite genomes and crossover.
+    int numElite = floor(elitePercentage * speciesSize);
+    int numDrop = floor((speciesSize - numElite) / 2) * 2;  // Make sure even number.
+    numElite = speciesSize - numDrop;
+    for (int i = numElite; i < speciesSize - 1; ++i)
+    {
+      // Randomly select two parents from elite genomes.
+      int idx1 = RandInt(0, numElite);
+      int idx2 = RandInt(0, numElite);
+
+      // Crossover to get two children genomes.
+      CrossoverWeights(species.genomes[idx1], species.genomes[idx2],
+                       species.genomes[i], species.genomes[i + 1]);
+    }
+
+    // Keep the best genome and mutate the rests.
+    for (int i = 1; i < speciesSize; ++i)
+    {
+      MutateWeightsBiased(species.genomes[i], mutateRate, mutateSize);
+    }
+  }
+
+  /**
+   * Evolution of species.
+   *
+   * The whole process of CNE algorithm.
+   * Iteratively evolve genomes for a number of generations to get a solution
+   * genome for a specific task.
+   */
+  bool Evolve()
+  {
+    // Generate initial species at random.
+    int generation = 0;
+    InitSpecies();
+    
+    // Repeat
+    while (generation < maxGeneration)
+    {
+    	// Evaluate all genomes in the species.
+      for (int i = 0; i < species.SpeciesSize(); ++i)
+      {
+        double fitness = task.EvalFitness(species.genomes[i]);
+        species.genomes[i].Fitness(fitness);
+      }
+      species.SetBestFitnessAndGenome();
+
+    	// Output some information.
+      printf("Generation: %zu\tBest fitness: %f\n", generation, species.BestFitness());
```

Comment:
Can you use ``Log::Info << "message" << std::endl;`` instead of ``printf``? That way a user can disable the output if needed or adjust the output according to his own preference e.g. only debug outputs (``Log::Debug``).